Set values not updated in state during Update with 0.12

### Terraform Version

```
Terraform master at rev f65b7c5 vendored into the Google provider (https://github.com/terraform-providers/terraform-provider-google/pull/2910)
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
resource "google_compute_ssl_policy" "update" {
	name = "name"
	description = "Generated by TF provider acceptance test"
	min_tls_version = "TLS_1_0"
	profile = "MODERN"
}
```

```hcl
resource "google_compute_ssl_policy" "update" {
	name = "name"
	description = "Generated by TF provider acceptance test"
	min_tls_version = "TLS_1_2"
	profile = "RESTRICTED"
}
```

Source: https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_compute_image_test.go#L270

### Debug Output
Relevant snippets:
```
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="44,43,39"
Cache-Control: private, max-age=0, must-revalidate, no-transform
Content-Type: application/json; charset=UTF-8
Date: Wed, 23 Jan 2019 22:53:45 GMT
Etag: "MU5Cd3oUlqkgEmH0MroUg5sNJ4I=/vvnLDHYivhhY7QZT-QZlvnJ-PS4="
Expires: Wed, 23 Jan 2019 22:53:45 GMT
Server: GSE
Vary: Origin
Vary: X-Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block

{
 "kind": "compute#sslPolicy",
 "id": "9149092770099842718",
 "creationTimestamp": "2019-01-23T14:53:37.369-08:00",
 "selfLink": "https://www.googleapis.com/compute/v1/projects/graphite-test-rileykarson/global/sslPolicies/test-ssl-policy-8dpyp9effa",
 "name": "test-ssl-policy-8dpyp9effa",
 "description": "Generated by TF provider acceptance test",
 "profile": "MODERN",
 "minTlsVersion": "TLS_1_0",
 "enabledFeatures": [
  "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
  "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
  "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA",
  "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
  "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
  "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA",
  "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
  "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
  "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
 ],
 "fingerprint": "KiUkvb-jIRc="
}
```


```
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="44,43,39"
Cache-Control: private, max-age=0, must-revalidate, no-transform
Content-Type: application/json; charset=UTF-8
Date: Wed, 23 Jan 2019 22:53:54 GMT
Etag: "0i3YPrxJAX2KiXLwWD9UizZqu1c=/6NF87P2VWM3VglSwe8YQhSsNIVQ="
Expires: Wed, 23 Jan 2019 22:53:54 GMT
Server: GSE
Vary: Origin
Vary: X-Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block

{
 "kind": "compute#sslPolicy",
 "id": "9149092770099842718",
 "creationTimestamp": "2019-01-23T14:53:37.369-08:00",
 "selfLink": "https://www.googleapis.com/compute/v1/projects/graphite-test-rileykarson/global/sslPolicies/test-ssl-policy-8dpyp9effa",
 "name": "test-ssl-policy-8dpyp9effa",
 "description": "Generated by TF provider acceptance test",
 "profile": "RESTRICTED",
 "minTlsVersion": "TLS_1_2",
 "enabledFeatures": [
  "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
  "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
  "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
  "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
  "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
 ],
 "fingerprint": "iKrWAkggR3o="
}

-----------------------------------------------------
```

```
--- FAIL: TestAccComputeSslPolicy_updateX (23.19s)
    testing.go:568: Step 3 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) (len=1) {
         (string) (len=18) "enabled_features.#": (string) (len=1) "6"
        }


        (map[string]string) (len=5) {
         (string) (len=18) "enabled_features.#": (string) (len=2) "10",
         (string) (len=27) "enabled_features.2321868343": (string) (len=34) "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
         (string) (len=27) "enabled_features.2631858259": (string) (len=36) "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA",
         (string) (len=26) "enabled_features.790781625": (string) (len=36) "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
         (string) (len=26) "enabled_features.966279389": (string) (len=34) "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
        }

FAIL
```

### Expected Behavior
The test succeeds (as it does on 0.11)

### Actual Behavior
The test fails; the state generated by updating the existing resource has 4 incorrect values, and the state from import is correct.

### Steps to Reproduce
```
make testacc TEST=./google TESTARGS='-parallel=1 -run=TestAccComputeSslPolicy_update'
```

### Additional Context
Running in CI with 0.12 SDK

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

I'm still seeing the same failures when I run against https://github.com/terraform-providers/terraform-provider-google/pull/2753, where we've vendored master @ `70eeec8`

```
------- Stdout: -------
=== RUN   TestAccComputeSslPolicy_update
--- FAIL: TestAccComputeSslPolicy_update (21.97s)
	testing.go:568: Step 3 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
		
		(map[string]string) (len=1) {
		 (string) (len=18) "enabled_features.#": (string) (len=1) "6"
		}
		
		
		(map[string]string) (len=5) {
		 (string) (len=18) "enabled_features.#": (string) (len=2) "10",
		 (string) (len=27) "enabled_features.2321868343": (string) (len=34) "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
		 (string) (len=27) "enabled_features.2631858259": (string) (len=36) "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA",
		 (string) (len=26) "enabled_features.790781625": (string) (len=36) "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
		 (string) (len=26) "enabled_features.966279389": (string) (len=34) "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
		}
		
FAIL
```

--- a/builtin/providers/test/provider.go
+++ b/builtin/providers/test/provider.go
@@ -30,6 +30,7 @@ func Provider() terraform.ResourceProvider {
 			"test_resource_deprecated":       testResourceDeprecated(),
 			"test_resource_defaults":         testResourceDefaults(),
 			"test_resource_list":             testResourceList(),
+			"test_resource_list_set":         testResourceListSet(),
 			"test_resource_map":              testResourceMap(),
 			"test_resource_computed_set":     testResourceComputedSet(),
 		},

--- a/builtin/providers/test/provider.go
+++ b/builtin/providers/test/provider.go
@@ -31,6 +31,7 @@ func Provider() terraform.ResourceProvider {
 			"test_resource_defaults":         testResourceDefaults(),
 			"test_resource_list":             testResourceList(),
 			"test_resource_map":              testResourceMap(),
+			"test_resource_computed_set":     testResourceComputedSet(),
 		},
 		DataSourcesMap: map[string]*schema.Resource{
 			"test_data_source":    testDataSource(),

--- a/builtin/providers/test/resource_computed_set.go
+++ b/builtin/providers/test/resource_computed_set.go
@@ -1,0 +1,74 @@
+package test
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func testResourceComputedSet() *schema.Resource {
+	return &schema.Resource{
+		Create: testResourceComputedSetCreate,
+		Read:   testResourceComputedSetRead,
+		Delete: testResourceComputedSetDelete,
+		Update: testResourceComputedSetUpdate,
+
+		CustomizeDiff: func(d *schema.ResourceDiff, _ interface{}) error {
+			o, n := d.GetChange("set_count")
+			if o != n {
+				d.SetNewComputed("string_set")
+			}
+			return nil
+		},
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"set_count": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"string_set": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Set: schema.HashString,
+			},
+		},
+	}
+}
+
+func testResourceComputedSetCreate(d *schema.ResourceData, meta interface{}) error {
+	d.SetId(fmt.Sprintf("%x", rand.Int63()))
+	return testResourceComputedSetRead(d, meta)
+}
+
+func testResourceComputedSetRead(d *schema.ResourceData, meta interface{}) error {
+	count := 3
+	v, ok := d.GetOk("set_count")
+	if ok {
+		count = v.(int)
+	}
+
+	var set []interface{}
+	for i := 0; i < count; i++ {
+		set = append(set, fmt.Sprintf("%d", i))
+	}
+
+	d.Set("string_set", schema.NewSet(schema.HashString, set))
+	return nil
+}
+
+func testResourceComputedSetUpdate(d *schema.ResourceData, meta interface{}) error {
+	return testResourceComputedSetRead(d, meta)
+}
+
+func testResourceComputedSetDelete(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("")
+	return nil
+}

--- a/builtin/providers/test/resource_computed_set_test.go
+++ b/builtin/providers/test/resource_computed_set_test.go
@@ -1,0 +1,52 @@
+package test
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestResourceComputedSet_update(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_computed_set" "foo" {
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"test_resource_computed_set.foo", "string_set.#", "3",
+					),
+				),
+			},
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_computed_set" "foo" {
+	set_count = 5
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"test_resource_computed_set.foo", "string_set.#", "5",
+					),
+				),
+			},
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_computed_set" "foo" {
+	set_count = 2
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"test_resource_computed_set.foo", "string_set.#", "2",
+					),
+				),
+			},
+		},
+	})
+}

--- a/builtin/providers/test/resource_list_set.go
+++ b/builtin/providers/test/resource_list_set.go
@@ -1,0 +1,73 @@
+package test
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func testResourceListSet() *schema.Resource {
+	return &schema.Resource{
+		Create: testResourceListSetCreate,
+		Read:   testResourceListSetRead,
+		Delete: testResourceListSetDelete,
+		Update: testResourceListSetUpdate,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"list": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"set": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"elem": {
+										Type:     schema.TypeString,
+										Optional: true,
+										DiffSuppressFunc: func(_, o, n string, _ *schema.ResourceData) bool {
+											return o == n
+										},
+									},
+								},
+							},
+							Set: func(v interface{}) int {
+								raw := v.(map[string]interface{})
+								if el, ok := raw["elem"]; ok {
+									return schema.HashString(el)
+								}
+								return 42
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func testResourceListSetCreate(d *schema.ResourceData, meta interface{}) error {
+	d.SetId(fmt.Sprintf("%x", rand.Int63()))
+	return testResourceListSetRead(d, meta)
+}
+
+func testResourceListSetUpdate(d *schema.ResourceData, meta interface{}) error {
+	return testResourceListSetRead(d, meta)
+}
+
+func testResourceListSetRead(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func testResourceListSetDelete(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("")
+	return nil
+}

--- a/builtin/providers/test/resource_list_set_test.go
+++ b/builtin/providers/test/resource_list_set_test.go
@@ -1,0 +1,55 @@
+package test
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestResourceListSet_basic(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_list_set" "foo" {
+  list {
+    set {
+      elem = "A"
+    }
+    set {
+      elem = "B"
+    }
+  }
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_list_set.foo", "list.0.set.1255198513.elem", "B"),
+					resource.TestCheckResourceAttr("test_resource_list_set.foo", "list.0.set.3554254475.elem", "A"),
+					resource.TestCheckResourceAttr("test_resource_list_set.foo", "list.0.set.#", "2"),
+				),
+			},
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_list_set" "foo" {
+  list {
+    set {
+      elem = "B"
+    }
+    set {
+      elem = "C"
+    }
+  }
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_list_set.foo", "list.0.set.1255198513.elem", "B"),
+					resource.TestCheckResourceAttr("test_resource_list_set.foo", "list.0.set.1037565863.elem", "C"),
+					resource.TestCheckResourceAttr("test_resource_list_set.foo", "list.0.set.#", "2"),
+				),
+			},
+		},
+	})
+}

--- a/builtin/providers/test/resource_nested_set_test.go
+++ b/builtin/providers/test/resource_nested_set_test.go
@@ -496,3 +496,34 @@ resource "test_resource_nested_set" "foo" {
 		},
 	})
 }
+
+func TestResourceNestedSet_multipleUnknownSetElements(t *testing.T) {
+	checkFunc := func(s *terraform.State) error {
+		return nil
+	}
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_nested_set" "a" {
+}
+
+resource "test_resource_nested_set" "b" {
+}
+
+resource "test_resource_nested_set" "c" {
+	multi {
+		optional = test_resource_nested_set.a.id
+	}
+	multi {
+		optional = test_resource_nested_set.b.id
+	}
+}
+				`),
+				Check: checkFunc,
+			},
+		},
+	})
+}

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -44,6 +44,10 @@ type GRPCProviderServer struct {
 }
 
 func (s *GRPCProviderServer) GetSchema(_ context.Context, req *proto.GetProviderSchema_Request) (*proto.GetProviderSchema_Response, error) {
+	// Here we are certain that the provider is being called through grpc, so
+	// make sure the feature flag for helper/schema is set
+	schema.SetProto5()
+
 	resp := &proto.GetProviderSchema_Response{
 		ResourceSchemas:   make(map[string]*proto.Schema),
 		DataSourceSchemas: make(map[string]*proto.Schema),

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -454,7 +454,6 @@ func (s *GRPCProviderServer) ReadResource(_ context.Context, req *proto.ReadReso
 	}
 
 	newStateVal = copyTimeoutValues(newStateVal, stateVal)
-	newStateVal = normalizeNullValues(newStateVal, stateVal, false)
 
 	newStateMP, err := msgpack.Marshal(newStateVal, block.ImpliedType())
 	if err != nil {

--- a/helper/schema/schema.go
+++ b/helper/schema/schema.go
@@ -19,6 +19,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"sync"
 
 	"github.com/hashicorp/terraform/config"
 	"github.com/hashicorp/terraform/terraform"
@@ -31,6 +32,27 @@ const PanicOnErr = "TF_SCHEMA_PANIC_ON_ERROR"
 
 // type used for schema package context keys
 type contextKey string
+
+var (
+	protoVersionMu sync.Mutex
+	protoVersion5  = false
+)
+
+func isProto5() bool {
+	protoVersionMu.Lock()
+	defer protoVersionMu.Unlock()
+	return protoVersion5
+
+}
+
+// SetProto5 enables a feature flag for any internal changes required required
+// to work with the new plugin protocol.  This should not be called by
+// provider.
+func SetProto5() {
+	protoVersionMu.Lock()
+	defer protoVersionMu.Unlock()
+	protoVersion5 = true
+}
 
 // Schema is used to describe the structure of a value.
 //

--- a/helper/schema/schema.go
+++ b/helper/schema/schema.go
@@ -799,10 +799,19 @@ func (m schemaMap) diff(
 	for attrK, attrV := range unsupressedDiff.Attributes {
 		switch rd := d.(type) {
 		case *ResourceData:
-			if schema.DiffSuppressFunc != nil &&
-				attrV != nil &&
+			if schema.DiffSuppressFunc != nil && attrV != nil &&
 				schema.DiffSuppressFunc(attrK, attrV.Old, attrV.New, rd) {
-				continue
+				// If this attr diff is suppressed, we may still need it in the
+				// overall diff if it's contained within a set. Rather than
+				// dropping the diff, make it a NOOP.
+				if !all {
+					continue
+				}
+
+				attrV = &terraform.ResourceAttrDiff{
+					Old: attrV.Old,
+					New: attrV.Old,
+				}
 			}
 		}
 		diff.Attributes[attrK] = attrV

--- a/helper/schema/set.go
+++ b/helper/schema/set.go
@@ -198,6 +198,16 @@ func (s *Set) add(item interface{}, computed bool) string {
 	code := s.hash(item)
 	if computed {
 		code = "~" + code
+
+		if isProto5() {
+			tmpCode := code
+			count := 0
+			for _, exists := s.m[tmpCode]; exists; _, exists = s.m[tmpCode] {
+				count++
+				tmpCode = fmt.Sprintf("%s%d", code, count)
+			}
+			code = tmpCode
+		}
 	}
 
 	if _, ok := s.m[code]; !ok {

